tests/wormhole/test_silicon_driver_wh.cpp

```diff
@@ -981,6 +981,11 @@ TEST(SiliconDriverWH, LargeAddressTlb) {
     // Read the scratch register via a different TLB, different code path:
     cluster.read_from_device(&value2, ARC_CORE, addr, sizeof(uint32_t), "REG_TLB");
 
+    // Mask off lower 16 bits; FW changes these dynamically:
+    value0 &= 0xffff0000;
```

Comment:
Can you elaborate (at least here in github comment), why is this expected? The chips are reset before running tests right? When they are reset firmware starts executing and I guess it comes to some stable state? Do you think that this exact register might be used as a scratch register and oscillating between values, which this write/read operation interferes with?

Reply:
I asked Andrew about the behavior of this register, I will reproduce my question and his answer:

Q: Can you explain the behavior of the register at 0x1ff30060 in Wormhole BAR0?  I have a unit test that reads it, and also reads it via TLB window to ARC.  In CI, there's intermittent failures because the BAR0 read doesn't match the TLB window read.  0xc0de0042 vs 0xc0de004d.  I don't want to mask off the low 16 bits without understanding why...

A: That's scratch 0, the "post code". The ARC FW writes status info there and it changes. A valid ARC FW post code has 0xc0de in the upper 16 bits. The low 16 bits may indicate where we are in the boot process or what the FW did last (e.g. handle a message). I'd say that 0x1 and 0x30-0x4F are normal codes.


This test wants to ensure that we are looking at the same thing regardless of whether we're using the REG_TLB path (dynamically assigned TLB window) or via a statically assigned TLB window.  I added this test while fixing addressing bugs that unnecessarily limited static TLB assignment and REG_TLB path to 32 bit addresses.  Since the NOC address of this register is > 32 bits, and because it is visible in BAR0, it is a good candidate for this sanity check.

Reply:
Does this mean that boot process advances exactly at the moment when you're reading the value in three different ways? It should eventually settle on a value, right?

Also, whatever is the answer, can we just read some other address without masking?